### ISSUE TITLE
Removing jQuery dependency -- A safe/sane approach

### ISSUE THREAD

**Issue by dannyrb** [cornerstonejs/cornerstone#180]
### We depend on jQuery for...

* Promises
* DOM selection (mostly in examples)
* Triggering custom events
* Listening for custom events
* HTTP calls?

### All of these can be replaced with vanilla JS equivalents (with a dash of polyfill) to:

* Maintain IE 10+ support
* Reduce overall filesize
* Eliminate a "global" dependency
* Leverage "single purpose" libraries

### I think the easiest way to transition is as follows:

**1. Add duplicate [custom event](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events) triggers in `Cornerstone` using vanilla js**
    - [Current Support](http://caniuse.com/#search=CustomEvent)
    - [Polyfill for older browsers](https://www.npmjs.com/package/custom-event-polyfill)

`Cornerstone` and `CornerstoneTools` will continue to work using the `jQuery` events, but this allows us to migrate existing tools in chunks, and to create new tools that rely on `custom event` without worrying about breaking existing functionality.

**2. Migrate tools and examples as we build out [the new documentation]**(https://dannyrb.gitbooks.io/cornerstone-tools/) to use vanilla DOM selection and the new `custom events`.

The tight coupling between `Cornerstone` and `CornerstoneTools` and what we use for events has likely been the biggest deterrent in dropping jQuery. The rest of the work is much easier, and satisfied with "up for grabs" issues.

**3. Clean up any remaining jQuery event based logic in either library.**

Now we no longer need to worry about the implications changes either library will have on the other. We can finish our jQuery cleanup in each as a separate initiative.

**4. At our leisure, replace any jQuery HTTP requests with [axios](https://github.com/axios/axios) equivalents.**

**5. Replace jQuery promises with [bluebird](https://github.com/petkaantonov/bluebird)**
    - [Support](https://github.com/petkaantonov/bluebird/issues/1429)

Throw a 🌲 on the 🔥 
Bust open some 🍾 
Because it's time to 🎉 🎈 

Goodbye jQuery

@swederik, curious to hear your thoughts. I think this can be broken down into a series of smaller issues, and then we can track the progress of this larger goal using GitHub milestones/projects?

Would love any feedback/suggestions you may have before I take a stab at item number 1 for this repository.

**Comment by swederik** [comment]
This sounds like a good approach to me as long as we're sure some people won't start picking up duplicate events. I assume that a jQuery listener won't detect a CustomEvent, but I've never checked...

Some work to switch to CustomEvents was already done by @zachasme here:
https://github.com/chafey/cornerstone/compare/zachasme-es6

Unfortunately it was during the ES6 transition so the branch is a bit of a mess, but the changes are there.

**Comment by dannyrb** [comment]
@swederik, I wish. I whipped up a quick code pen to show that jQuery catches vanilla js custom events:

https://codepen.io/dannyrb/pen/jGKvjN/

We can account for this by **not** reusing the same event names as we slowly implement this. Which we can quickly rename using a quick find-and-replace after step 3 is completed and we've officially removed all event code

The trick is to not advertise the new events as interfaces/spec that should be coded against. Any code using the current event naming will continue to work throughout this whole process. Win-win.

Let me know if you have questions, I may not have explained that well.

@zachasme, given the changes that have taken place since your PR, would you be offended if we started this fresh and incrementally?

**Comment by zachasme** [comment]
@dannyrb: No offense taken---starting fresh is the way to go. I've applied my Promise-related changes on another branch: https://github.com/chafey/cornerstone/compare/master...zachasme:removejquery

For the tests I've used the Promise polyfill that babel uses. I would advise against using a library that adds "sugar" on top of Promises (e.g. bluebird), considering this issue is partly due code dependant on jQuerys Deferred.

As for transitioning to custom events, if you cant reuse event names, you could go for the lowercase naming scheme used for native event types ('mousedown', 'loadedmetadata', etc.).

Glad to see jQuery go! 👍

**Comment by dannyrb** [comment]
@zachasme, appreciate the thoughts on promises/libs. I also like the native event type naming scheme. Mixed case naming prevents the event binding features in Vue from working (:

What are your thoughts on excluding polyfills from the module, and noting helpful polyfills to gain additional support? Here's a SO answer that got me thinking: https://stackoverflow.com/a/28965677/1867984

I do like your [polyfill solution here](https://github.com/chafey/cornerstone/pull/97/files#diff-56f863957d69c401148521727309d6a9) -- It's small enough that it might not be worth the trouble of pointing would be `cornerstone` users toward an "optional" polyfill they will most likely need.

@swederik, maybe I'm missing something, but why do we trigger events against `cornerstone.events` instead of DOM element in some instances?

**Comment by zachasme** [comment]
I agree a Promise polyfill should not be bundled with the distribution. Documenting it as an environment requirement and perhaps adding a runtime check is a good approach. But, as you say, the CustomEvent polyfill is small enough it could probably be bundled.

Also do note that CustomEvents can only be fired on DOM-elements, so you will need a simple event listener implementation such as:
https://github.com/zachasme/cornerstone/blob/es6/src/addEventListener.js

**Comment by zachasme** [comment]
I've pushed a possible implementation to https://github.com/chafey/cornerstone/compare/master...zachasme:removejquery.

It adds a `createAndDispatchEvent` helper that triggers a CustomEvent as well as a jQuery event, using a mapping from new to old event names. This way only two lines need be removed in order to deprecate jQuery support.

**Comment by dannyrb** [comment]
I really like your implementation @zachasme. Is the thought that `cornerstone.events.addEventListener` would only be used internally to listen to events on `cornerstone.events`? I'm 👌 with that, but if they're meant to be available externally ( @swederik ), then wouldn't it be better to have them dispatch off the `enabledElement`?

@swederik, how do you feel about this as solution to part 1 listed above?
> 1. Add duplicate custom event triggers in Cornerstone using vanilla js

All existing jQuery events would continue to exist (for now). And it allows us to proceed with steps 2-5. The biggest caveat is:

* Once we completely remove jQuery, we'll be using new event names
    * Breaking change, but the new names allow for easier listening by other popular frameworks:
        * Vue: https://github.com/vuejs/vue/issues/4839

Curious to hear your thoughts @swederik && @zachasme.

**Comment by chafey** [comment]
@dannyrb & @zachasme thank you for pushing this forward.  It has been a long time desire of mine to remove jQuery as a dependency so I am happy to see it go!  Your approach looks good, but any breaking changes (like even renaming) needs to have a very strong reason.  I have not used Vue.js but am not inclined to change the event naming just to make it easier for the few users using it at the expense of the thousands of users not.  The exception to this is if there is now a well accepted standard for custom event naming (I am not aware of this) and we want to conform to it

**Comment by swederik** [comment]
The thing is that some of those events are not specific to any enabled element (e.g. image load progress), so it's not obvious where on the DOM they should be attached. WADO Image Loader & Cornerstone Tools also both use events triggered on cornerstone.events (previously $(cornerstone), but modules broke that). I am open to suggestions.

I don't mind if we have to change the names as long as there is clear documented migration steps and it's done for a major version so nobody accidentally updates and breaks their entire app.

I agree with both of you to keep polyfills out of these libraries, and agree with @zachasme that we should remove the dependency altogether, not just replace it with bluebird or something else.

On my side, I've started migrating from jQuery Deferred to Promises for Cornerstone and WADO Image Loader and it seems to be working. Basically following @diego0020 and @leonardorame's plan here: https://github.com/chafey/cornerstone/issues/93

Image loaders in this branch return { promise: a standard promise, cancelFn: A function to cancel the request? haven't implemented it yet}. loadImage and loadAndCacheImage will only return the promise. Let me know if this makes sense.

See:
https://github.com/chafey/cornerstone/tree/noDeferred
https://github.com/chafey/cornerstoneWADOImageLoader/tree/noDeferred

**Comment by dannyrb** [comment]
> I don't mind if we have to change the names as long as there is clear documented migration steps and it's done for a major version so nobody accidentally updates and breaks their entire app. - @swederik 

> Your approach looks good, but any breaking changes (like even renaming) needs to have a very strong reason. - @chafey 

Different schools of thought ;) But we can conform to either of these. My thought/hope is that most users depend on a specific version of `cornerstone` for production instances. Breaking changes are a part of major version upgrades, and migration docs need to be followed -- and this is a particularly small breaking change for a major version bump.

The only valid-ish reasons I have for a name change are:

- `CustomEvent`'s Event.type is case insensitive: [MSDN Docs](https://developer.mozilla.org/en-US/docs/Web/API/Event/type)
- Native event naming looks like: `mouseout`, `click`, `dblclick`

I agree that we shouldn't change for any single framework, but I could see us transitioning names to:

- Match native event naming: `CornerstoneImageLoaded` --> `cornerstoneimageloaded`
- Kebab case: `CornerstoneImageLoaded` --> `cornerstone-image-loaded`

Bike-shedding a little bit here, so I'll assume for now that we're going to keep the same event names unless I hear otherwise from @chafey or @swederik. I'd like to see the change (it'd make this transition a little easier), but the end goal is to 🔪 ☠️ jQuery with 🔥 , and I don't want to see event naming get in the way of that (:

Thanks for your explanation of `cornerstone.events` @swederik! To make sure I've got it right, it's best to think of `cornerstone.events` as global events that aren't specific to an `enabledElement`? Really, it makes sense to think about these events as "framework level" events that we [publish and others can subscribe](https://davidwalsh.name/pubsub-javascript) to? (not dissimilar from @zachasme's implementation)

If that's the case, then we need to make sure we clearly document how to subscribe to these events, and maybe work to expose slightly friendlier methods for those subscriptions.

**Comment by diego0020** [comment]
I would be happy to help in this effort. Let me know if you need anything

**Comment by Maistho** [comment]
Regarding the naming, I've seen other libraries scope all their events with `<library>:<event>` which would make the events look like `cornerstone:newimageloaded`. Seems like a good way to make it clear that the events are cornerstone-specific, and if cornerstoneTools wanted to add events only for cornerstoneTools they could use `cornerstonetools:eventname` to make it clearer where the event is actually coming from.

**Comment by Maistho** [comment]
Sidenote regarding `cornerstone.events`, which is not a `HTMLElement`. It doesn't have any event target logic attached to it. When using custom events with `dispatchEvent`, there is some more work needed.

A solution could be to just trigger the events on `document`, with namespacing (`cornerstone:imageloaded`) and no bubbling. To listen for those events one would `document.addEventListener('cornerstone:imageloaded', myListener)`.

I think it might be a better approach to make `cornerstone.events` implement the [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) interface instead, but I'm open to ideas.

**Comment by Maistho** [comment]
I'll also note that there is less breaking changes by making `cornerstone.events` implement the EventTarget interface, since the jQuery listeners will continue to work in the same way as previous.

**Comment by swederik** [comment]
I've added the EventTarget interface to cornerstone.events, merged the PR that triggers parallel CustomEvents and released Version 0.13.2. You are now free to listen on the CustomEvents instead of the jQuery Events.

I've also updated https://github.com/chafey/cornerstoneWADOImageLoader to 0.15.1 and made the same changes. It requires Cornerstone Core 0.13.2 since the events are dispatched on cornerstone.events.

Next are Tools and the Web Image Loader...

**Comment by Maistho** [comment]
Great work @swederik!

I've actually done some work on Tools already in a private branch, I'll make sure to open a PR towards your branch later today.

I'll  also take another pass at the triggerEvent function. I realised after looking at @zachasme #97 that it's a lot better to move all the jquery.trigger into that function as well.

Should event names be changed to lowercase, or remain CamelCase?

**Comment by dannyrb** [comment]
@swederik, from what I can tell, you and a few others have already taken a large chunk out of this issue. Is it possible to setup a milestone that shows which steps still need to be completed, are in progress, and need claimed? Anything in particular that is still "up for grabs"?

All of this effort is amazing, but I want to make sure we don't end up in a position where everyone thinks someone else is taking care of X.

**Comment by swederik** [comment]
OK so here goes..

Core and the image loaders are basically only waiting on the $.Deferred -> Promise transition. 

The problem is that if we update Core now, Tools will not be compatible with it. So I think we should release at least a minor version of Tools that can use the CustomEvents emitted from Core before we drop jQuery from Core (as suggested in number 2. from the first post by @dannyrb).

## Step 1: Minor version of Cornerstone Tools

I would really appreciate help with this. @Maistho has already stepped up which is great, but it will require a lot of testing to verify that current functionality isn't broken.

- [ ] Consume Custom Events from Cornerstone Core instead of jQuery events (I've started this here: https://github.com/chafey/cornerstoneTools/tree/custom-events)
- [ ] Trigger custom events in addition to jQuery events for events that library consumers might be using (@maistho has started this here: https://github.com/chafey/cornerstoneTools/pull/274)
- [ ] Switch Magnify tool to CustomEvent: 

Magnify tool adds additional information when binding to CornerstoneNewImage which native event binding doesn't support. Need to work around this. https://github.com/chafey/cornerstoneTools/blob/e1d2c19c27ec11dd49bc7f8980bdae975c74ba93/src/imageTools/magnify.js#L42

## Step 2: Major versions of Cornerstone Core & Image Loaders

- [ ] Settle on what image loaders should return (#184). **Review the pull request**.

We currently return $.Deferred from image loaders. #184 migrates this to an Object containing a Promise and an optional function used to cancel the request. If I don't get any more comments after Step 1 is done, I am going to assume it's good enough and just move forward.

- [ ] Remove jQuery from the triggerEvent function, start using the lower-case names exclusively.

This would mean that all applications listening for these events would stop working.

### Make related breaking change for image loaders, release major versions:
- [ ] Update WADO Image Loader to return the new image loader object instead of a Deferred (https://github.com/chafey/cornerstoneWADOImageLoader/pull/137)
- [ ] Update Web Image Loader to return the new image loader object (not done yet). The Deferred here needs to be removed: https://github.com/chafey/cornerstoneWebImageLoader/blob/master/src/loadImage.js#L16
- [ ] Remove jQuery events from WADO Image Loader (e.g. https://github.com/chafey/cornerstoneWADOImageLoader/blob/0825fd0f38ac17c34629293066c5482c17615cb3/src/imageLoader/internal/xhrRequest.js#L38)
- [ ] Remove jQuery events from Web Image Loader (https://github.com/chafey/cornerstoneWebImageLoader/blob/master/src/loadImage.js#L51)

## Step 3: Major version of Cornerstone Tools (2.0.0)

- [ ] Drop jQuery events entirely, switch all tools to use custom events (this will be a long, tedious process)
- [ ] Write instructions for users with custom tools to transition their event handlers (e.g users of mouseButtonTool)

Thoughts?

**Comment by Maistho** [comment]
I've started converting some mouse tools already, since I still have issues with jQuery event propagation not working properly.  (I think I still have multiple instances of jQuery for some reason)

Can confirm that converting tools is a long tedious process. It's mostly quite straightforward though.

**Comment by swederik** [comment]
I think we really should try to split the minor / major updates to Cornerstone Tools into two pieces, or all the other changes will be waiting for that work to be done. I have some other obligations in the way but I will try to get back to the Cornerstone Tools custom-events branch ASAP.

**Comment by dannyrb** [comment]
Thanks for pointing out specific branches and ongoing PRs for step 1's minor patch. 

While we're doing this, are there any tests/formats you'd specifically like to see? Are there any examples you can point would-be-contributors toward? Hard to add tests after the fact, but if we're in the middle of a refactor. *shrugs*

Thanks again for kickstarting this step and providing good example refactors for a few tools. I'll add what I can this weekend and early next week 👍

**Comment by swederik** [comment]
I think I am done with Step 1 here: https://github.com/chafey/cornerstoneTools/pull/288

I'd be happy for some code reviews to make sure I didn't miss anything. If anyone has the time to run through all the examples and make sure they still work, that would also be great. Live examples for the custom-events branch is here: https://rawgit.com/chafey/cornerstoneTools/custom-events/examples/index.html

**Comment by swederik** [comment]
Just released Cornerstone Tools 1.1.0 which should finish Step 1.

**Comment by swederik** [comment]
Finally released Cornerstone Core, WADO Image Loader, and Web Image Loader major versions without jQuery! 

Thanks to all that have helped! Please test and report any issues with these libraries. Only Tools is left, so I would appreciate any PRs to help make Cornerstone jQuery-free!

**Comment by swederik** [comment]
Anyone that feels like testing Cornerstone Tools can try this PR:

https://github.com/cornerstonejs/cornerstoneTools/pull/303

and test these examples live:

https://rawgit.com/cornerstonejs/cornerstoneTools/v2/examples/index.html

Please let me know of any issues so we can fix them!

**Comment by swederik** [comment]
Done! Closing this beast :-)
